Introduce `secrets.properties`

wzieba opened the pull request:
## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR splits `gradle.properties` into 3 files:
- `gradle.properties` which constains only build related configuration and is now checked in into VCS
- `secrets.properties` which contains only secrets, fetched from `mobile-secrets` tool (with accompanying `defaults.properties`)
- `developer.properties-example` which allows to set local, custom developer configuration

These changes make the setup safer by putting secrets outside of the working repository and easier to provide build configuration changes, as no `mobile-secrets` update is requested to get them anymore. It also streamlines setting up the project as there's no need to copy `gradle.properties-example` to build the project.

-----

`dependency-tree-diff` step won't succeed because we changed the build steps comparing to `trunk` (removed copying `gradle.properties-example`)

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

### Signing configuration

1. Checkout `trunk` and setup all secrets `bundle exec fastlane run configure_apply`. 
2. Run `:WooCommerce:signingReport` task, copy the content
3. Checkout this branch, run `:WooCommerce:signingReport` again
4. Compare outputs of both executions. They should be the same.

Repeat these steps for `WooCommerce-Wear` module.

### App secrets

Open generated `BuildConfig.kt` file. Make sure it contains valid secrets. You can compare content of `BuildConfig.kt` from build from `trunk`, but be aware they'll be different: unused secrets have been removed from mobile secrets.

### Measure builds

Apply secrets `bundle exec fastlane run configure_apply` and assert, that at the end of the task ` Build time report of Xm Ys has been received by Apps Metrics.` is printed to the console.

### Developer properties

1. Copy `developer.properties-example`
2. Provide a change in the configuration
3. Assert that the change of property is reflected in generated `BuildConfig.kt` file.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Request `secrets.properties` from mobile secrets

wzieba committed:
Load secrets from `secrets.properties`. Introduce `developer.properties` for local developer settings

wzieba committed:
Apply secrets to `measureBuilds` extension

wzieba committed:
Make `signingConfigs` extension use secrets from `secrets.properties`

wzieba committed:
Include `gradle.properties` file (with build configuration only), remove `gradle.properties-example`

Remove mentioning `reset_db_on_downgrade` as this property is not used anywhere

wzieba committed:
Remove unused `checkGradlePropertiesFile` method

wzieba committed:
Move `IAP_TESTING_SANDBOX_URL` and `JITM_TESTING_JSON_FILE_NAME` to `developer.properties`

It's appropriate file for developer, device-specific configuration

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba commented on .configure:
```diff
@@ -1,11 +1,11 @@
   "project_name": "woocommerce-android",
-  "branch": "trunk",
-  "pinned_hash": "126ffacbbfc7abf1f4bf14670d72418846f04673",
+  "branch": "wcandroid_secrets_split",
```
- [x] Before merge of this PR,  `wcandroid_secrets_split` needs to be merged to `trunk` and this file needs to be updated

ParaskP7 replied:
Noticing your comment here @wzieba , and after reviewing this PR, I am still unsure as to why this was done. Comparing to how it was done for JP/WPAndroid last week ([PR](https://github.com/wordpress-mobile/WordPress-Android/pull/21442)), I am unsure of the need to do it different this time. I did notice this `wcandroid_secrets_split` being created on the `.mobile-secrets` repo. Can you help me understand the need for this extra step? 🙏

wzieba replied:
I wasn't sure how long this PR will take to review and then address review comments, so I didn't want to push these changes to `trunk` in mobile-secrets in case someone wanted to update secrets in old `gradle.properties`. I risked pushing changes directly to `trunk` in WPAndroid, as development there has a focus on different repository and I didn't thought there's a lot of chance they will modify secrets soon.

ParaskP7 replied:
Got it, thanks for the clarification on that Wojtek, much appreciated, makes sense and thanks for taking steps to ensure smooth migration! 💯

wzieba commented on WooCommerce/build.gradle:
```diff
@@ -523,14 +544,6 @@ android.buildTypes.all { buildType ->
         }
     }
 
-    buildConfigField "String",
-            "IAP_TESTING_SANDBOX_URL",
-            "\"${project.properties.getOrDefault('iap_testing_sandbox_url', '')}\""
-
-    buildConfigField "String",
-            "JITM_TESTING_JSON_FILE_NAME",
-            "\"${project.properties.getOrDefault('jitm_testing_json_file_name', '')}\""
```
@kidinov I moved these properties to `developers.properties-example` file to unify how this project sets up developer-only local configuration. This file will work similarly to how `gradle.properties-example` worked - `developer.properties` is not checked into VCS.

I don't how you and the team uses this property, but does this solution work for you? One thing to note is that passing these properties through command line won't work now.

kidinov replied:
👋 @wzieba 

That's fine! We use it rarely but still use

wzieba committed:
Update fastlane-plugin-wpmreleasetoolkit to 12.3.2

To address an issue with initial pull of secrets

ParaskP7 reviewed (commented):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, (almost) everything works as expected, what an awesome job you did here too, so beautiful, so much more organized, thank YOU! 🌟 x 🌟 ^ 🌟 

FYI: I added this (almost) parenthesis above because I can't see to be able to see `Measure Builds` running, or else, printing the `Build time report ...` at the end of a task. Can you please verify for me that this works as expected. Then, it is a ✅ for me. 🙏 

PS: As far as the `Build Config` diff is concerned, although everything looks good, the only change I noticed is these 2 `IAP_TESTING_SANDBOX_URL` and `JITM_TESTING_JSON_FILE_NAME` strings being added to `WooCommerce-Wear` module, which previously only existed on the `WooCommerce` module. But, based on the restructure, I understand this is expected, and I don't think of them as being harmful. I only wanted to notice it here, just in case.

wzieba committed:
Use secrets properties when determining if publish Gradle scan

wzieba committed:
Use shared value of `userEnabledTracking` to protect against forgetting to update this value in one place

wzieba commented:
Thank you @ParaskP7 🙌 . Indeed the measure builds stopped working because I've missed updating how the project gets `tracksEnabled`. I've addressed the problem and extracted the value to a shared value, to reduce risk of similar mistake in the future.

wzieba committed:
Rename `tracksEnabled` to `measureBuildsEnabled`
Tracks is also name for user interaction tracking, so 'tracksEnabled' might be confusing

ParaskP7 commented:
👋 @wzieba and thanks for [this](https://github.com/woocommerce/woocommerce-android/pull/12933#issuecomment-2483811254) update, quite weird and sneaky that this `userEnabledTracking` where within this `gradle_build_scan.gradle` file (I guess a leftover), thanks for moving it out of there (one more reason to migrate away of Groovy and using such patterns and onto Gradle Kotlin DSL)! 🥇 

I testing this again and everything now works as expected, 🟢 light from me! 🚀

wzieba committed:
Update configure to point to trunk
